zkevm-circuits/src/util/int_decomposition.rs

```diff
@@ -0,0 +1,100 @@
+//! Define IntDecomposition to decompose int into byte limbs
+use eth_types::{Field, ToLittleEndian, H160, U256};
+use gadgets::util::{sum, Expr};
+use halo2_proofs::{
+    circuit::{AssignedCell, Value},
+    plonk::{Error, Expression},
+};
+use itertools::Itertools;
+
+use crate::evm_circuit::{
+    param::N_BYTES_HALF_WORD,
+    util::{rlc, CachedRegion, Cell},
+};
+
+use super::word::{Word, WordExpr};
+
+#[derive(Clone, Debug)]
+/// IntDecomposition decompose integer into byte limbs
+pub struct IntDecomposition<F, const N_LIMBS: usize> {
+    /// inner cells in little-endian for synthesis
+    pub limbs: [Cell<F>; N_LIMBS],
+}
+
+impl<F: Field, const N_LIMBS: usize> IntDecomposition<F, N_LIMBS> {
+    /// new by cell limbs
+    pub fn new(limbs: [Cell<F>; N_LIMBS]) -> Self {
+        Self { limbs }
+    }
+
+    /// assign bytes to cells
+    pub fn assign<const N_LIMBS_ASSIGN: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes: Option<[u8; N_LIMBS_ASSIGN]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        assert!(N_LIMBS_ASSIGN >= N_LIMBS);
```

Comment:
Shouldn't this be `N_LIMBS >= N_LIMBS_ASSIGN`?
Otherwise there are more bytes to assign than limbs, so the more significant bytes are ignored.

Reply:
More significant bytes ignored is expected because we have so many place to put smaller value in u256 so here we allow larger bytes to smaller limbs, which also make caller side easiler

Reply:
Hm, I see.  I think it would be better if we only assign types that can fit (and if we have an address in u256, have the caller convert it to address before assigning).  But we can leave this discussion for the future.

Nevertheless, to follow this current implementation I would ask for the following safety check:
for each byte that is not assigned, assert it is 0.
Something like this:
```rust
for byte in bytes[N_LIMBS_ASSIGN-N_LIMBS..] {
    assert_eq(byte, 0);
}
```

Reply:
Really nice suggestion! Added check in this commit https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1411/commits/f245af8248fa073011e0c0d7675c00a206b4da53